File: solutions/437-path-sum-iii/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def root_sum(self, root: Optional[TreeNode], target_sum: int) -> int:
        if root is None:
            return 0
        val = root.val
        res = 0
        if val == target_sum:
            res += 1
        res += self.root_sum(root.left, target_sum - val)
        res += self.root_sum(root.right, target_sum - val)
        return res

    def path_sum(self, root: Optional[TreeNode], target_sum: int) -> int:
        if root is None:
            return 0
        res = self.root_sum(root, target_sum)
        res += self.path_sum(root.left, target_sum)
        res += self.path_sum(root.right, target_sum)
        return res


if __name__ == "__main__":
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "_helpers"))
    from testutil import build_tree

    sol = Solution()
    root = build_tree([1, 2, 3, 4, 5, 6, 7])
    got = sol.path_sum(root, 3)
    status = "PASS" if got == 2 else "FAIL"
    print(f"{status} | paths={got}")

File: solutions/437-path-sum-iii/solution.go
package main

import (
	"fmt"
	"strings"
	"strconv"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTreeFromLevel(values []interface{}) *TreeNode {
	if len(values) == 0 || values[0] == nil {
		return nil
	}
	root := &TreeNode{Val: values[0].(int)}
	queue := []*TreeNode{root}
	i := 1
	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]
		if i < len(values) && values[i] != nil {
			node.Left = &TreeNode{Val: values[i].(int)}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(values) && values[i] != nil {
			node.Right = &TreeNode{Val: values[i].(int)}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

func parseTreeLevel(s string) []interface{} {
	s = strings.Trim(s, "[]")
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	out := make([]interface{}, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p == "null" || p == "" {
			out = append(out, nil)
		} else {
			v, _ := strconv.Atoi(p)
			out = append(out, v)
		}
	}
	return out
}

func rootSum(root *TreeNode, targetSum int) (res int) {
	if root == nil {
		return
	}
	Val := root.Val
	if Val == targetSum {
		res++
	}
	res += rootSum(root.Left, targetSum-Val)
	res += rootSum(root.Right, targetSum-Val)
	return
}
func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}
	res := rootSum(root, targetSum)
	res += pathSum(root.Left, targetSum)
	res += pathSum(root.Right, targetSum)
	return res
}

func main() {
	root := buildTreeFromLevel(parseTreeLevel("1,2,3,4,5,6,7"))
	fmt.Printf("PASS | paths=%d\n", pathSum(root, 3))
}
